### PR TITLE
Refactor Tty.

### PR DESCRIPTION
**Pull request by reitermarkus**
Restored some of the code from Cask's now removed `Tty` module and added a `Formatter` module to help with re-using predefined styles.

### CODE REVIEW

File: Library/Homebrew/cask/lib/hbc/cli/style.rb [hunk 1]
```diff
@@ -28,7 +28,8 @@ def install_rubocop
       begin
         Homebrew.install_gem_setup_path! "rubocop-cask", RUBOCOP_CASK_VERSION, "rubocop"
       rescue SystemExit
-        raise Hbc::CaskError, $stderr.string.chomp.sub("#{Tty.red}Error#{Tty.reset}: ", "")
+        require "utils/tty"
+        raise Hbc::CaskError, $stderr.string.chomp.extend(Tty::StringExtension).strip_ansi.gsub("Error: ", "")
```
Comment (jawshooah) [inline Library/Homebrew/cask/lib/hbc/cli/style.rb]: I think it's much cleaner to either make `strip_ansi` a module method or just monkeypatch `String`. It doesn't make sense to extend each individual string just to call this method.

Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/cli/style.rb]: Reverted it to being a module method.

File: Library/Homebrew/utils.rb [hunk 2]
```diff
@@ -10,102 +10,28 @@
...
     title = Tty.truncate(title)
   end
-  puts "#{Tty.green}==>#{Tty.white} #{title}#{Tty.reset}"
+  puts "#{Tty.green}==>#{Tty.reset} #{Tty.bold}#{title}#{Tty.reset}"
 end
 
 # Print a warning (do this rarely)
 def opoo(warning)
-  $stderr.puts "#{Tty.yellow}Warning#{Tty.reset}: #{warning}"
+  $stderr.puts "#{Tty.yellow.underline}Warnin#{Tty.reset.yellow}g:#{Tty.reset} #{warning}"
```
Comment (chdiza) [inline Library/Homebrew/utils.rb]: Looks like the word "Warning" got broken up.

Comment (reitermarkus) [inline Library/Homebrew/utils.rb]: This is intentional, so the `g` doesn't get underlined.

Comment (chdiza) [inline Library/Homebrew/utils.rb]: If you don't mind me asking, why is underlining being added here where there was none before?

Comment (reitermarkus) [inline Library/Homebrew/utils.rb]: It's not being added. Before, [`yellow`](https://github.com/Homebrew/brew/pull/823/files#diff-cef16ceb727f1d5a841e0295b083611fL36) was actually a shortcut to `yellow.underline`.

Comment (jawshooah) [inline Library/Homebrew/utils.rb]: Wait, why are we underlining everything but the "g"?

Comment (reitermarkus) [inline Library/Homebrew/utils.rb]: Because the underline would collide with the `g`. This is really a minor detail. 😄

Comment (jawshooah) [inline Library/Homebrew/utils.rb]: That it is, just a weird one. But it makes sense now 👍

Comment (UniqMartin) [inline Library/Homebrew/utils.rb]: > Because the underline would collide with the `g`. This is really a minor detail. 😄

I don't understand this. Would you mind explaining it in a little more detail? Thanks!

**Comment by UniqMartin** [comment]
Thanks for taking a stab at this, but I think this goes into the wrong direction and further reinforces a not so good style. My personal opinion is that we should make the styling more semantic rather than hard-coding the style on every use (think of templates rather than direct formatting in word processing). Examples:

- `#{Tty.red.underline}Error#{Tty.reset}` should be `#{Tty.error}Error#{Tty.reset}` and the exact styling (currently red and underlined) an implementation detail of the `Tty.error` method.

  I think it would be nice to also style the following colon in this and similar cases (warning, …), but that's a different debate.

- `#{Tty.underline}URL#{Tty.reset}` (previously `#{Tty.em}URL#{Tty.reset}`) should be `#{Tty.url}URL#{Tty.reset}`.

Not only will this make styling more consistent, but it will also expose when we have too many different styles and the output becomes more colorful and varied than useful. If we can keep the number of `Tty.<predefined_style>` methods limited, this will come almost for free.

> *No description provided.*

I appreciate your work a lot, but I would be a bit happier if you could provide some description of your PRs. Mainly a why and some high-level overview would be nice to have, as those aren't immediately obvious from looking at the diff. This doesn't have to be long, but the length should be non-zero except for small self-explanatory or urgent fixes.

File: Library/Homebrew/utils.rb (hunk 2 above)
Comment (reitermarkus) [inline Library/Homebrew/utils.rb]: @UniqMartin, it's just that from a design point-of-view, the underline should not clash with the descenders of lowercase letters.

**Comment by reitermarkus** [comment]
> we should make the styling more semantic

I agree, but speaking of semantics, formatting shouldn't be part of the `Tty` module.

File: Library/Homebrew/utils.rb (hunk 2 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/utils.rb]: I think the underline should remain in that case or not be used in words with `g`s

Comment (MikeMcQuaid) [inline Library/Homebrew/utils.rb]: As this is Homebrew's case: I think it should remain as it was.

Comment (MikeMcQuaid) [inline Library/Homebrew/utils.rb]: (the previous version looks fine/good to me)

File: Library/Homebrew/utils/tty.rb [hunk 3]
```diff
@@ -0,0 +1,45 @@
+module Tty
+  module_function
+
+  def strip_ansi(str)
+    str.gsub(/\033\[\d+(;\d+)*m/, "")
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/tty.rb]: `str` -> `string`

File: Library/Homebrew/utils/tty.rb [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  end
+
+  def width
+    `/usr/bin/tput cols`.strip.to_i
+  end
+
+  def truncate(str)
+    (w = width).zero? ? str.to_s : str.to_s[0, w - 4]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/tty.rb]: `str` -> `string`, use a longer variable for `w`, use `if` over ternary, use a more self-documenting function for `[]`

File: Library/Homebrew/utils/tty.rb [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    magenta:       35,
+    cyan:          36,
+    default:       39,
+
+    # Styles
+    reset:          0,
+    bold:           1,
+    italic:         3,
+    underline:      4,
+    strikethrough:  9,
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/tty.rb]: Don't think it's worth aligning these.

File: Library/Homebrew/utils/tty.rb [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  }.each do |name, code|
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/tty.rb]: I'd consider declaring this as two frozen Hash constants and iterating over that; makes it a big more explicit what's going on.

File: Library/Homebrew/utils/tty.rb [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    define_singleton_method(name) do
+      @sequence = [*@sequence, code]
+      self
+    end
+  end
+
+  def to_s
+    return "" if [*@sequence].empty? || !$stdout.tty?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/tty.rb]: Split into two `return`s

File: Library/Homebrew/utils/tty.rb [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    "\033[#{@sequence.join(";")}m"
+  ensure
+    @sequence = []
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/tty.rb]: This state feels kinda weird to me.

**Comment by MikeMcQuaid** [comment]
> #{Tty.underline}URL#{Tty.reset} (previously #{Tty.em}URL#{Tty.reset}) should be #{Tty.url}URL#{Tty.reset}.

I like this.

> I appreciate your work a lot, but I would be a bit happier if you could provide some description of your PRs.

👍 


> I agree, but speaking of semantics, formatting shouldn't be part of the Tty module.

I think we can consider renaming the module.

The metaprogramming here is fine but I think it's worth being extremely explicit with variable naming and splitting statements onto more lines so it's more obvious to the casual, non-expert Rubyist what's going on here.

File: Library/Homebrew/cask/lib/hbc/artifact/moved.rb [hunk 9]
```diff
@@ -79,10 +79,9 @@ def summarize_artifact(artifact_spec)
     if target.exist?
       target_abv = " (#{target.abv})"
     else
-      warning = "Missing #{self.class.artifact_english_name}"
-      warning = "#{Tty.red}#{warning}#{Tty.reset}: "
+      error = "#{Tty.red}Missing #{self.class.artifact_english_name}:#{Tty.reset} "
```
Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/artifact/moved.rb]: Note to self: `Formatter.error`

**Comment by reitermarkus** [comment]
@UniqMartin or @MikeMcQuaid, could someone take a look again?

File: Library/Homebrew/utils/tty.rb [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+  end
+
+  COLOR_CODES = {
+                  red:     31,
+                  green:   32,
+                  yellow:  33,
+                  blue:    34,
+                  magenta: 35,
+                  cyan:    36,
+                  default: 39,
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/tty.rb]: Tiny nit: just indent these at four spaces in and use hash rockets for now (we'll do a mass update 🔜)

File: Library/Homebrew/utils/tty.rb [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+                }
+
+  STYLE_CODES = {
+                  reset:          0,
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/tty.rb]: Similarly here.

File: Library/Homebrew/utils/tty.rb [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+                  underline:      4,
+                  strikethrough:  9,
+                  no_underline:  24,
+                }
+
+  CODES = COLOR_CODES.merge(STYLE_CODES)
+
+  CODES.each do |name, code|
+    define_singleton_method(name) do
+      @sequence = [*@sequence, code]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/tty.rb]: This is the bit I'm not sure I understand what/why it's doing what it is.

File: Library/Homebrew/utils/tty.rb [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+      self
+    end
+  end
+
+  def to_s
+    return "" if [*@sequence].empty?
+    return "" unless $stdout.tty?
+    "\033[#{@sequence.join(";")}m"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/tty.rb]: Maybe worth making the `\033` a constant somewhere; used in a couple of places.

**Comment by MikeMcQuaid** [comment]
Looking much better, thanks @reitermarkus.

File: Library/Homebrew/utils/tty.rb (hunk 12 above)
Comment (reitermarkus) [inline Library/Homebrew/utils/tty.rb]: Since `@sequence` can be `nil` (initially), we simply use the splat operator to ensure that the end result is always an array which consists of the current `@sequence` followed by the `code`.

File: Library/Homebrew/utils/tty.rb (hunk 10 above)
Comment (reitermarkus) [inline Library/Homebrew/utils/tty.rb]: Why *four* spaces, not two?

File: Library/Homebrew/utils/tty.rb (hunk 13 above)
Comment (reitermarkus) [inline Library/Homebrew/utils/tty.rb]: The only other place this is used is in the tests. And I'd rather not use constants that are defined inside the to-be-tested class.

File: Library/Homebrew/utils/tty.rb (hunk 10 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/tty.rb]: Just because it's in a module hence 2+2.

File: Library/Homebrew/utils/tty.rb (hunk 12 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/tty.rb]: Sorry, higher-level: why do we need a `@sequence` variable now and didn't before?

File: Library/Homebrew/utils/tty.rb (hunk 13 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/tty.rb]: Used in `strip_ansi` above?

File: Library/Homebrew/utils/tty.rb (hunk 12 above)
Comment (reitermarkus) [inline Library/Homebrew/utils/tty.rb]: So we can use chaining like `Tty.red.underline`.

Comment (MikeMcQuaid) [inline Library/Homebrew/utils/tty.rb]: It feels like this might be done nicer by returning actual instances?

Comment (MikeMcQuaid) [inline Library/Homebrew/utils/tty.rb]: Or if not: can we split it into more methods to make what's going on more obvious.

File: Library/Homebrew/utils/tty.rb (hunk 13 above)
Comment (reitermarkus) [inline Library/Homebrew/utils/tty.rb]: Ah, I missed that. However, how would I do this? The regex needs a literal `'\033\['`, while the `to_s` method needs the escaped `"\033\["`.

File: Library/Homebrew/utils/tty.rb (hunk 12 above)
Comment (reitermarkus) [inline Library/Homebrew/utils/tty.rb]: I think returning actual instances would only complicate this. And what do you want to split into more methods here? There's only one `each` loop defining methods, how do you split an `each` loop?

File: Library/Homebrew/utils/formatter.rb [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+    else
+      "#{Tty.send(color).underline}#{descenders_no_underline(label)}#{Tty.reset}#{message}"
+    end
+  end
+
+  def descenders_no_underline(string)
+    colon_reached = false
+
+    string.chars.map { |letter|
+      unless colon_reached
```
Comment (jawshooah) [inline Library/Homebrew/utils/formatter.rb]: Was this left in by accident? Doesn't the first line of `colorize_text_before_colon` take care of this?

Comment (reitermarkus) [inline Library/Homebrew/utils/formatter.rb]: Yes, good catch.

File: Library/Homebrew/utils/tty.rb (hunk 13 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/tty.rb]: Ah I see, fine as-is then 👍

File: Library/Homebrew/utils/tty.rb (hunk 12 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/tty.rb]: I don't know, really. I just find the current implementation confusing; it's not obvious that it's for changing and how the `@sequence` stuff works.

Some vague thoughts:
```ruby
@sequence ||= []
@sequence << code
self
```
signals intent a bit more. Even then, I'd be tempted to make a module-level method that is called `append_code` that encapsulates this logic so the `define_singleton_method` just ends up calling it. Similarly with `[*@sequence].empty?` below; having that be a separate module-level method would be good. Finally, using an attribute reader for `sequence` seems better than having raw access given how it's intended to be used.

Comment (MikeMcQuaid) [inline Library/Homebrew/utils/tty.rb]: I still think it would be good to split this into separate intent-signaling methods. I have my head around the `@sequence` stuff now but it's still a lot more complex to understand from the previous code. My preference is self-documenting code by splitting things into multiple methods that explain clearly what they do through their names. If you'd rather not do that: please add a lot more comments explaining what this is doing, how and why. Thanks!

File: Library/Homebrew/dev-cmd/bottle.rb [hunk 15]
```diff
@@ -92,7 +93,8 @@ def print_filename(string, filename)
       next unless ARGV.verbose? && !text_matches.empty?
       print_filename string, file
       text_matches.first(MAXIMUM_STRING_MATCHES).each do |match, offset|
-        puts " #{Tty.gray}-->#{Tty.reset} match '#{match}' at offset #{Tty.em}0x#{offset}#{Tty.reset}"
+        puts " #{Tty.gray}-->#{Tty.reset} match '#{match}' at offset #{Tty.underline}0x#{offset}#{Tty.reset}"
+
```
Comment (UniqMartin) [inline Library/Homebrew/dev-cmd/bottle.rb]: This looks like a superfluous empty line to me.

File: Library/Homebrew/formula_installer.rb [hunk 16]
```diff
@@ -212,7 +212,7 @@ def install
       opoo "#{formula.full_name}: #{old_flag} was deprecated; using #{new_flag} instead!"
     end
 
-    oh1 "Installing #{Tty.green}#{formula.full_name}#{Tty.reset}" if show_header?
+    oh1 "Installing #{Formatter.success(formula.full_name)}" if show_header?
```
Comment (UniqMartin) [inline Library/Homebrew/formula_installer.rb]: Using `Formatter.success` in this and several other contexts below where it encloses a formula/dependency name feels somewhat strange to me. Should there be a dedicated formatter for these cases?

File: Library/Homebrew/test/.simplecov [hunk 17]
```diff
@@ -9,11 +9,11 @@ SimpleCov.start do
...
-  add_filter "/Homebrew/cask/spec/"
-  add_filter "/Homebrew/cask/test/"
-  add_filter "/Homebrew/compat/"
-  add_filter "/Homebrew/test/"
-  add_filter "/Homebrew/vendor/"
+  add_filter "/Library/Homebrew/cask/spec/"
+  add_filter "/Library/Homebrew/cask/test/"
+  add_filter "/Library/Homebrew/compat/"
+  add_filter "/Library/Homebrew/test/"
+  add_filter "/Library/Homebrew/vendor/"
```
Comment (UniqMartin) [inline Library/Homebrew/test/.simplecov]: Why the change? In any case, this feels misplaced as this is completely unrelated to the changes in this PR (unless I'm missing something).

File: Library/Homebrew/dev-cmd/test-bot.rb [hunk 18]
```diff
@@ -178,19 +178,19 @@ def puts_command
         puts "travis_fold:start:#{@travis_fold_id}"
         puts "travis_time:start:#{@travis_timer_id}"
       end
-      puts "#{Tty.blue}==>#{Tty.white} #{@command.join(" ")}#{Tty.reset}"
+      puts Formatter.headline(@command.join(" "), color: :blue)
```
Comment (UniqMartin) [inline Library/Homebrew/dev-cmd/test-bot.rb]: I don't feel too strongly about this, but my impression is we want to avoid introducing the new hash syntax until we update our RuboCop configuration and make a sweeping change, thus `:color => :blue` here and similarly in other places.

File: Library/Homebrew/utils/formatter.rb [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+
+    if label.nil?
+      "#{Tty.send(color)}#{message}#{Tty.reset}"
+    else
+      "#{Tty.send(color).underline}#{descenders_no_underline(label)}#{Tty.reset}#{message}"
+    end
+  end
+
+  def descenders_no_underline(string)
+    string.gsub(/([-+:jgqpy])/, "#{Tty.no_underline}\\1#{Tty.underline}")
```
Comment (UniqMartin) [inline Library/Homebrew/utils/formatter.rb]: Why is this excluding `-`, `+`, and `:` that are clearly characters that stay above the baseline? Think about the `:` in `Formatter.url("http://example.com")`; not underlining this colon looks really weird.

Aside from that, I'm really against this kind of micro-optimization, not to mention that it only looks decent for letters where the descender spans the width of the entire character, e.g. `g` in most fonts, but looks like something is missing for letters like `p`. The problem with the underline is not that it touches descenders but the underline itself that is undesirable from a typographical point of view. Furthermore, this manual treatment will defeat proper treatment by the font and/or the font rendering of my terminal (assuming it did have some clever handling of underlines like a word processor often has).

Finally, check out e.g. the `Security` link on the bottom of every GitHub page. If you hover over it, the whole text is underlined (including the `y`). Like everyone else, we shouldn't bother special-casing this.

Practical suggestions:

- Underline all characters of a URL unconditionally.
- Underline all characters of an error/warning label unconditionally or simply avoid underlining if this feels like it looks too ugly (and yes, underlines are always ugly).

File: Library/Homebrew/utils/formatter.rb [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+
+  def error(string)
+    colorize_text_before_colon(string, :red)
+  end
+
+  def url(string)
+    descenders_no_underline("#{Tty.blue.underline}#{string}#{Tty.reset}")
+  end
+
+  def colorize_text_before_colon(string, color)
```
Comment (UniqMartin) [inline Library/Homebrew/utils/formatter.rb]: Just a nit pick, but this actually colorizes the text before the colon *and* the colon itself (which IMO is good from a stylistic point of view), so the method name isn't 100% spot-on.

File: Library/Homebrew/utils/tty.rb [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+      self
+    end
+  end
+
+  def to_s
+    return "" if @sequence.nil?
+    return "" unless $stdout.tty?
+    "\033[#{@sequence.join(";")}m"
+  ensure
+    @sequence = nil
```
Comment (UniqMartin) [inline Library/Homebrew/utils/tty.rb]: Should we optimize for the next invocation and initialize this to an empty array instead? I guess the downside would be that this method would need to check for both `@sequence.nil?` and `@sequence.empty?` in its first line, though I that's probably only going to be the case for `"#{Tty}"` or `Tty.to_s` which would be bogus anyway, right?

File: Library/Homebrew/formula_installer.rb (hunk 16 above)
Comment (reitermarkus) [inline Library/Homebrew/formula_installer.rb]: I agree. I only changed this so it would show up in the diff, so we could discuss a better name or format.

File: Library/Homebrew/test/.simplecov (hunk 17 above)
Comment (reitermarkus) [inline Library/Homebrew/test/.simplecov]: Yes, this is unrelated, somehow got in from working on the coverage PR.

File: Library/Homebrew/utils/formatter.rb (hunk 20 above)
Comment (reitermarkus) [inline Library/Homebrew/utils/formatter.rb]: `colorize_label`, maybe?

File: Library/Homebrew/utils/formatter.rb (hunk 19 above)
Comment (reitermarkus) [inline Library/Homebrew/utils/formatter.rb]: I can live with your practical suggestions. 👍

File: Library/Homebrew/utils/tty.rb (hunk 21 above)
Comment (reitermarkus) [inline Library/Homebrew/utils/tty.rb]: That's exactly why I had a check for `[*@sequence].empty?` here before, which would have included both cases.

File: Library/Homebrew/dev-cmd/bottle.rb [hunk 22]
```diff
@@ -70,6 +70,7 @@ def print_filename(string, filename)
         print_filename(string, file) unless linked_libraries.empty?
         linked_libraries.each do |lib|
           puts " #{Tty.gray}-->#{Tty.reset} links to #{lib}"
+
```
Comment (reitermarkus) [inline Library/Homebrew/dev-cmd/bottle.rb]: All these empty lines are so the places where `Tty` is used instead of a `Formatter` show up in the diff. I wasn't sure if these warrant a `Formatter`, or what to name it.

File: Library/Homebrew/debrew.rb [hunk 23]
```diff
@@ -111,7 +111,7 @@ def self.debug(e)
 
     begin
       puts "#{e.backtrace.first}"
-      puts "#{Tty.red}#{e.class.name}#{Tty.reset}: #{e}"
+      puts Formatter.error("#{e.class.name}: #{e}")
```
Comment (reitermarkus) [inline Library/Homebrew/debrew.rb]: I suppose we could do something like `Formatter.error(e, label: e.class.name)`.

Comment (UniqMartin) [inline Library/Homebrew/debrew.rb]: Judging from what I've seen in the code, `Formatter.error([<label>, ]<message>)` would be more practical, as I don't think it makes sense to omit “label” (or more precisely the thing that should be styled). This is how it would be used and what it would yield:

```ruby
Formatter.error("Custom label", "Message") # => "_Custom label:_ Message"
Formatter.error("Error message") # => "_Error:_ Error message"
```

Here, `_<whatever>_` denotes the styled part. Would that work?

File: Library/Homebrew/dev-cmd/bottle.rb (hunk 22 above)
Comment (UniqMartin) [inline Library/Homebrew/dev-cmd/bottle.rb]: I see; this makes sense. I didn't realize this and (if mentioned) overlooked it in previous comments. I think it's important to take our first few steps in the right direction (like with the error message and URL formatting), but we don't necessarily have to eliminate all direct uses of `Tty` in this PR, though we should consider doing so in the not too distant future. But as usual, naming is hard as is identifying the necessary styles while consolidating the probably too many different styles currently in use.

File: Library/Homebrew/formula_installer.rb (hunk 16 above)
Comment (UniqMartin) [inline Library/Homebrew/formula_installer.rb]: How about `Formatter.identifier` or `Formatter.name`?

File: Library/Homebrew/utils/formatter.rb (hunk 20 above)
Comment (UniqMartin) [inline Library/Homebrew/utils/formatter.rb]: Yep, `colorize_label` or possibly a more opaque `style_label` would have worked. (I guess that method name is no longer needed given the changes you pushed since then.)

File: Library/Homebrew/utils/tty.rb (hunk 21 above)
Comment (UniqMartin) [inline Library/Homebrew/utils/tty.rb]: I have to agree with Mike that I find the splatting and collecting a bit harder to follow. A possibly easier to read and still succinct alternative for the check could be:

```ruby
return "" if (@sequence || []).empty?
```

File: Library/Homebrew/brew.rb [hunk 24]
```diff
@@ -144,8 +144,8 @@ def require?(path)
   Utils::Analytics.report_exception(e)
   onoe e
   if internal_cmd && defined?(OS::ISSUES_URL)
-    $stderr.puts "#{Tty.white}Please report this bug:"
-    $stderr.puts "    #{Tty.em}#{OS::ISSUES_URL}#{Tty.reset}"
+    $stderr.puts "Please report this bug:"
+    $stderr.puts "    #{Formatter.url(OS::ISSUES_URL)}"
```
Comment (UniqMartin) [inline Library/Homebrew/brew.rb]: I think I'd like to preserve the boldface in the first line (`#{Tty.bold}Please report this bug:#{Tty.reset}`) and, since we're touching this anyway, adjust the indentation of the second line to two spaces which better aligns with most (if not all) similar user-visible output.

File: Library/Homebrew/dev-cmd/bottle.rb (hunk 15 above)
Comment (UniqMartin) [inline Library/Homebrew/dev-cmd/bottle.rb]: Also, in this specific case I would prefer to swap the underline for something else. `Tty.bold` would be fine by me, but I'm definitely open to alternative suggestions. It's just that the underline seriously and unnecessarily degrades legibility of those hex numbers.

File: Library/Homebrew/diagnostic.rb [hunk 25]
```diff
@@ -477,7 +477,7 @@ def check_user_curlrc
           If you have trouble downloading packages with Homebrew, then maybe this
           is the problem? If the following command doesn't work, then try removing
           your curlrc:
-            curl https://github.com
+            curl #{Formatter.url("https://github.com")}
```
Comment (UniqMartin) [inline Library/Homebrew/diagnostic.rb]: I would personally not bother highlighting this URL, as it's a command line argument to be copy-pasted together with the `curl` rather than a URL we expect users to check out in isolation (most, if not all, other instances where `Formatter.url` is used). But I don't feel overly strongly about this.

File: Library/Homebrew/formula_installer.rb [hunk 26]
```diff
@@ -421,7 +421,7 @@ def install_dependencies(deps)
     if deps.empty? && only_deps?
       puts "All dependencies for #{formula.full_name} are satisfied."
     elsif !deps.empty?
-      oh1 "Installing dependencies for #{formula.full_name}: #{Tty.green}#{deps.map(&:first)*", "}#{Tty.reset}",
+      oh1 "Installing dependencies for #{formula.full_name}: #{Formatter.success(deps.map(&:first)*", ")}",
```
Comment (UniqMartin) [inline Library/Homebrew/formula_installer.rb]: IMO, it would be nicer to highlight the individual elements and keep the delimiting comma and space unstyled instead of styling the entire string only after joining the array elements.

File: Library/Homebrew/utils/formatter.rb [hunk 27]
```diff
@@ -0,0 +1,46 @@
...
+
+  def success(string, label: nil)
+    label(label, string, :green)
+  end
+
+  def warning(string, label: nil)
+    label(label, string, :yellow)
+  end
+
+  def error(string, label: nil)
```
Comment (UniqMartin) [inline Library/Homebrew/utils/formatter.rb]: Sorry if I'm making you go back and forth a bit, but after giving this some more thought, I think the combined handling of label and message string just complicates things without providing sufficient benefit. I think it would be best to keep the concatenation of label and message string on the call site and let `Formatter.error` et al. just wrap the given string like other `Formatter` methods. I.e. it should be used as follows:

```ruby
Formatter.error("FAILED")
"#{Formatter.error("Error:")} #{message}"
"#{Formatter.error("READ THIS:")} #{Formatter.url("https://example.com/")}"
```

File: Library/Homebrew/utils/formatter.rb [hunk 28]
```diff
@@ -0,0 +1,46 @@
...
+    label(label, string, :red)
+  end
+
+  def url(string)
+    "#{Tty.blue.underline}#{string}#{Tty.reset}"
```
Comment (UniqMartin) [inline Library/Homebrew/utils/formatter.rb]: Would personally prefer to stick with the underline only (no color), but I don't feel very strongly about this.

File: Library/Homebrew/debrew.rb (hunk 23 above)
Comment (UniqMartin) [inline Library/Homebrew/debrew.rb]: Please ignore my most recent comment; it doesn't make much sense after reviewing the code once more. https://github.com/Homebrew/brew/pull/823#r79279639 is a much saner suggestion.

File: Library/Homebrew/utils/formatter.rb (hunk 27 above)
Comment (reitermarkus) [inline Library/Homebrew/utils/formatter.rb]: I think this would be a step back, as this gets us to basically start *formatting* on the call side again, which kinda defeats the purpose of having a formatter in the first place.

Comment (UniqMartin) [inline Library/Homebrew/utils/formatter.rb]: I kinda agree. How do you suggest we proceed? The following works for me:

```ruby
Formatter.error(message) # default label "Error"
Formatter.error(message, label: "READ THIS") # custom label
```

But I don't know how to deal with the case where there's neither a label nor a message, but just a string that should be styled like the error label. Should that be a separate method?

Comment (reitermarkus) [inline Library/Homebrew/utils/formatter.rb]: Without a label, the message will be styled the same as the label

`Formatter.error("This string is red.)`

`Formatter.error("This message is not red.", label: "Red Label")`.

File: Library/Homebrew/utils/tty.rb (hunk 21 above)
Comment (reitermarkus) [inline Library/Homebrew/utils/tty.rb]: I think we can leave this as-is. I don't even think *optimizing* would be worth it, since you will alway lose the last `[]`.

**Comment by MikeMcQuaid** [comment]
@reitermarkus Can you try to avoid pushing quite as much? Every time you do there's a CI job that takes ~30m to run.

**Comment by reitermarkus** [comment]
@MikeMcQuaid, sorry about that.

**Comment by MikeMcQuaid** [comment]
@reitermarkus No worries, not to know 👍

File: Library/Homebrew/utils/tty.rb (hunk 21 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/tty.rb]: I still think I'd really like pretty much all usage of `@sequence` to be in methods explaining what they do. I can take a stab at that if you don't mind me committing on this PR.

Comment (reitermarkus) [inline Library/Homebrew/utils/tty.rb]: > I can take a stab at that if you don't mind me committing on this PR.

Please do, I don't mind at all.

**Comment by ilovezfs** [comment]
:-1:

File: Library/Homebrew/dev-cmd/test-bot.rb (hunk 18 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/test-bot.rb]: > I don't feel too strongly about this, but my impression is we want to avoid introducing the new hash syntax until we update our RuboCop configuration and make a sweeping change

This was done.

File: Library/Homebrew/diagnostic.rb (hunk 25 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/diagnostic.rb]: I'm fine with highlighting it.

File: Library/Homebrew/utils/formatter.rb (hunk 27 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/formatter.rb]: I like the current version as-is.

File: Library/Homebrew/utils/formatter.rb (hunk 28 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/formatter.rb]: Actually, no, let's leave it colourless for now.

**Comment by MikeMcQuaid** [comment]
@reitermarkus Sorry it took me so long to get to this. I've updated `utils/tty.rb` in a way that I think signals intent a bit more and should make it more obvious what's going on. If you're 👍 on that and address the other comments (basically now just: don't colour URLs) and fix the merge conflicts (😭) then I'm happy for this to 🚢

**Comment by reitermarkus** [comment]
> fix the merge conflicts (:sob:) 

Totally looking forward to that! :joy:

**Comment by reitermarkus** [comment]
Have looked over this on more time. Merging this now to avoid any more merge conflicts.

**Comment by ilovezfs** [comment]
This seems to have broken CI.
https://bot.brew.sh/job/Homebrew%20Core%20Pull%20Requests/8588/version=el_capitan/console
See https://github.com/Homebrew/brew/issues/1209
